Windows shared library build problem under both VS2019 and MSYS2

I am not able to build the shared library under either VS2019 or MSYS2. Static build is okay.
Ref: https://github.com/avrdudes/avrdude/issues/952#issuecomment-1120810691
Ref: https://github.com/avrdudes/avrdude/issues/951#issuecomment-1120991143

Config log for VS2019.

```
1> CMake generation started for configuration: 'x64'.
1> Command line: "C:\WINDOWS\system32\cmd.exe" /c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Release" -DCMAKE_INSTALL_PREFIX:PATH="C:\work\avr\avrdude\out\install\x64" -DUSE_EXTERNAL:BOOL="True" -DCMAKE_BUILD_TYPE:STRING="Release" -DBUILD_SHARED_LIBS:BOOL="True"  -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" "C:\work\avr\avrdude" 2>&1"
1> Working directory: C:\work\avr\avrdude\out\build\x64
1> [CMake] -- Fetching external libraries, please wait...
1> [CMake] -- Using external library 'libelf'
1> [CMake] -- Using external library 'libusb'
1> [CMake] -- Using external library 'libhidapi'
1> [CMake] -- Using external library 'libftdi'
1> [CMake] -- Configuration summary:
1> [CMake] -- ----------------------
1> [CMake] -- DO HAVE    libelf
1> [CMake] -- DO HAVE    libusb
1> [CMake] -- DON'T HAVE libusb_1_0
1> [CMake] -- DO HAVE    libhidapi
1> [CMake] -- DO HAVE    libftdi
1> [CMake] -- DON'T HAVE libftdi1
1> [CMake] -- DISABLED   doc
1> [CMake] -- DISABLED   parport
1> [CMake] -- DISABLED   linuxgpio
1> [CMake] -- DISABLED   linuxspi
1> [CMake] -- ----------------------
1> [CMake] -- Configuring done
1> [CMake] -- Generating done
1> [CMake] -- Build files have been written to: C:/work/avr/avrdude/out/build/x64
1> Extracted CMake variables.
1> Extracted source files and headers.
1> Extracted code model.
1> Extracted toolchain configurations.
1> Extracted includes paths.
1> CMake generation finished.
```

Failed build log.

<details>
<summary> Click for full build log </summary>

```
>------ Build All started: Project: avrdude, Configuration: x64 ------
  [1/56] Building C object src\CMakeFiles\libavrdude.dir\dfu.c.obj
  [2/56] Building C object src\CMakeFiles\libavrdude.dir\avr910.c.obj
  [3/56] Building C object src\CMakeFiles\libavrdude.dir\arduino.c.obj
  [4/56] Building C object src\CMakeFiles\libavrdude.dir\fileio.c.obj
  [5/56] Building C object src\CMakeFiles\libavrdude.dir\avr.c.obj
  [6/56] Building C object src\CMakeFiles\libavrdude.dir\avrftdi.c.obj
  [7/56] Building C object src\CMakeFiles\libavrdude.dir\avrftdi_tpi.c.obj
  [8/56] Building C object src\CMakeFiles\libavrdude.dir\avrpart.c.obj
  [9/56] Building C object src\CMakeFiles\libavrdude.dir\buspirate.c.obj
  [10/56] Building C object src\CMakeFiles\libavrdude.dir\crc16.c.obj
  [11/56] Building C object src\CMakeFiles\libavrdude.dir\config.c.obj
  [12/56] Building C object src\CMakeFiles\libavrdude.dir\bitbang.c.obj
  [13/56] Building C object src\CMakeFiles\libavrdude.dir\butterfly.c.obj
  [14/56] Building C object src\CMakeFiles\libavrdude.dir\flip2.c.obj
  [15/56] Building C object src\CMakeFiles\libavrdude.dir\jtagmkI.c.obj
  [16/56] Building C object src\CMakeFiles\libavrdude.dir\ppi.c.obj
  [17/56] Building C object src\CMakeFiles\libavrdude.dir\serbb_posix.c.obj
  [18/56] Building C object src\CMakeFiles\libavrdude.dir\jtag3.c.obj
  [19/56] Building C object src\CMakeFiles\libavrdude.dir\jtagmkII.c.obj
  [20/56] Building C object src\CMakeFiles\libavrdude.dir\ft245r.c.obj
  [21/56] Building C object src\CMakeFiles\libavrdude.dir\flip1.c.obj
  [22/56] Building C object src\CMakeFiles\libavrdude.dir\linuxgpio.c.obj
  [23/56] Building C object src\CMakeFiles\libavrdude.dir\ppiwin.c.obj
  [24/56] Building C object src\CMakeFiles\libavrdude.dir\linuxspi.c.obj
  [25/56] Building C object src\CMakeFiles\libavrdude.dir\ser_avrdoper.c.obj
  [26/56] Building C object src\CMakeFiles\libavrdude.dir\lists.c.obj
  [27/56] Building C object src\CMakeFiles\libavrdude.dir\pgm.c.obj
  [28/56] Building C object src\CMakeFiles\libavrdude.dir\pgm_type.c.obj
  [29/56] Building C object src\CMakeFiles\libavrdude.dir\micronucleus.c.obj
  [30/56] Building C object src\CMakeFiles\libavrdude.dir\par.c.obj
  [31/56] Building C object src\CMakeFiles\libavrdude.dir\ser_posix.c.obj
  [32/56] Building C object src\CMakeFiles\libavrdude.dir\pindefs.c.obj
  [33/56] Building C object src\CMakeFiles\libavrdude.dir\confwin.c.obj
  [34/56] Building C object src\CMakeFiles\libavrdude.dir\stk500.c.obj
  [35/56] Building C object src\CMakeFiles\libavrdude.dir\stk500generic.c.obj
  [36/56] Building C object src\CMakeFiles\libavrdude.dir\updi_link.c.obj
  [37/56] Building C object src\CMakeFiles\libavrdude.dir\serialupdi.c.obj
  [38/56] Building C object src\CMakeFiles\libavrdude.dir\updi_nvm.c.obj
  [39/56] Building C object src\CMakeFiles\libavrdude.dir\updi_state.c.obj
  [40/56] Building C object src\CMakeFiles\libavrdude.dir\stk500v2.c.obj
  [41/56] Building C object src\CMakeFiles\libavrdude.dir\updi_readwrite.c.obj
  [42/56] Building C object src\CMakeFiles\libavrdude.dir\serbb_win32.c.obj
  [43/56] Building C object src\CMakeFiles\libavrdude.dir\teensy.c.obj
  [44/56] Building C object src\CMakeFiles\libavrdude.dir\wiring.c.obj
  [45/56] Building C object src\CMakeFiles\libavrdude.dir\usbasp.c.obj
  [46/56] Building C object src\CMakeFiles\libavrdude.dir\xbee.c.obj
  [47/56] Building C object src\CMakeFiles\libavrdude.dir\usb_hidapi.c.obj
  [48/56] Building C object src\CMakeFiles\libavrdude.dir\update.c.obj
  [49/56] Building C object src\CMakeFiles\libavrdude.dir\usbtiny.c.obj
  [50/56] Building C object src\CMakeFiles\libavrdude.dir\usb_libusb.c.obj
  [51/56] Building C object src\CMakeFiles\libavrdude.dir\ser_win32.c.obj
  [52/56] Building C object src\CMakeFiles\libavrdude.dir\pickit2.c.obj
  [53/56] Building C object src\CMakeFiles\libavrdude.dir\__\lexer.c.obj
  [54/56] Building C object src\CMakeFiles\libavrdude.dir\__\config_gram.c.obj
  [55/56] Linking CXX shared library src\libavrdude.dll
  FAILED: src/libavrdude.dll src/libavrdude.lib 
  cmd.exe /C "cd . && "C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_dll --intdir=src\CMakeFiles\libavrdude.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1431~1.311\bin\Hostx64\x64\link.exe /nologo src\CMakeFiles\libavrdude.dir\arduino.c.obj src\CMakeFiles\libavrdude.dir\avr.c.obj src\CMakeFiles\libavrdude.dir\avr910.c.obj src\CMakeFiles\libavrdude.dir\avrftdi.c.obj src\CMakeFiles\libavrdude.dir\avrftdi_tpi.c.obj src\CMakeFiles\libavrdude.dir\avrpart.c.obj src\CMakeFiles\libavrdude.dir\bitbang.c.obj src\CMakeFiles\libavrdude.dir\buspirate.c.obj src\CMakeFiles\libavrdude.dir\butterfly.c.obj src\CMakeFiles\libavrdude.dir\config.c.obj src\CMakeFiles\libavrdude.dir\confwin.c.obj src\CMakeFiles\libavrdude.dir\crc16.c.obj src\CMakeFiles\libavrdude.dir\dfu.c.obj src\CMakeFiles\libavrdude.dir\fileio.c.obj src\CMakeFiles\libavrdude.dir\flip1.c.obj src\CMakeFiles\libavrdude.dir\flip2.c.obj src\CMakeFiles\libavrdude.dir\ft245r.c.obj src\CMakeFiles\libavrdude.dir\jtagmkI.c.obj src\CMakeFiles\libavrdude.dir\jtagmkII.c.obj src\CMakeFiles\libavrdude.dir\jtag3.c.obj src\CMakeFiles\libavrdude.dir\linuxgpio.c.obj src\CMakeFiles\libavrdude.dir\linuxspi.c.obj src\CMakeFiles\libavrdude.dir\lists.c.obj src\CMakeFiles\libavrdude.dir\micronucleus.c.obj src\CMakeFiles\libavrdude.dir\par.c.obj src\CMakeFiles\libavrdude.dir\pgm.c.obj src\CMakeFiles\libavrdude.dir\pgm_type.c.obj src\CMakeFiles\libavrdude.dir\pickit2.c.obj src\CMakeFiles\libavrdude.dir\pindefs.c.obj src\CMakeFiles\libavrdude.dir\ppi.c.obj src\CMakeFiles\libavrdude.dir\ppiwin.c.obj src\CMakeFiles\libavrdude.dir\serbb_posix.c.obj src\CMakeFiles\libavrdude.dir\serbb_win32.c.obj src\CMakeFiles\libavrdude.dir\ser_avrdoper.c.obj src\CMakeFiles\libavrdude.dir\ser_posix.c.obj src\CMakeFiles\libavrdude.dir\ser_win32.c.obj src\CMakeFiles\libavrdude.dir\serialupdi.c.obj src\CMakeFiles\libavrdude.dir\stk500.c.obj src\CMakeFiles\libavrdude.dir\stk500v2.c.obj src\CMakeFiles\libavrdude.dir\stk500generic.c.obj src\CMakeFiles\libavrdude.dir\teensy.c.obj src\CMakeFiles\libavrdude.dir\updi_link.c.obj src\CMakeFiles\libavrdude.dir\updi_nvm.c.obj src\CMakeFiles\libavrdude.dir\updi_readwrite.c.obj src\CMakeFiles\libavrdude.dir\updi_state.c.obj src\CMakeFiles\libavrdude.dir\usbasp.c.obj src\CMakeFiles\libavrdude.dir\usb_hidapi.c.obj src\CMakeFiles\libavrdude.dir\usb_libusb.c.obj src\CMakeFiles\libavrdude.dir\usbtiny.c.obj src\CMakeFiles\libavrdude.dir\update.c.obj src\CMakeFiles\libavrdude.dir\wiring.c.obj src\CMakeFiles\libavrdude.dir\xbee.c.obj src\CMakeFiles\libavrdude.dir\__\lexer.c.obj src\CMakeFiles\libavrdude.dir\__\config_gram.c.obj  /out:src\libavrdude.dll /implib:src\libavrdude.lib /pdb:src\libavrdude.pdb /dll /version:1.0 /machine:x64 /INCREMENTAL:NO  _deps\libelf-build\libelf.lib  _deps\libusb-build\libusb.lib  "C:\Program Files (x86)\Windows Kits\10\Lib\10.0.22000.0\um\x64\hid.lib"  _deps\libhidapi-build\libhidapi.lib  _deps\libftdi-build\libftdi.lib  setupapi.lib  ws2_32.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib  && cd ."
  LINK: command "C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1431~1.311\bin\Hostx64\x64\link.exe /nologo src\CMakeFiles\libavrdude.dir\arduino.c.obj src\CMakeFiles\libavrdude.dir\avr.c.obj src\CMakeFiles\libavrdude.dir\avr910.c.obj src\CMakeFiles\libavrdude.dir\avrftdi.c.obj src\CMakeFiles\libavrdude.dir\avrftdi_tpi.c.obj src\CMakeFiles\libavrdude.dir\avrpart.c.obj src\CMakeFiles\libavrdude.dir\bitbang.c.obj src\CMakeFiles\libavrdude.dir\buspirate.c.obj src\CMakeFiles\libavrdude.dir\butterfly.c.obj src\CMakeFiles\libavrdude.dir\config.c.obj src\CMakeFiles\libavrdude.dir\confwin.c.obj src\CMakeFiles\libavrdude.dir\crc16.c.obj src\CMakeFiles\libavrdude.dir\dfu.c.obj src\CMakeFiles\libavrdude.dir\fileio.c.obj src\CMakeFiles\libavrdude.dir\flip1.c.obj src\CMakeFiles\libavrdude.dir\flip2.c.obj src\CMakeFiles\libavrdude.dir\ft245r.c.obj src\CMakeFiles\libavrdude.dir\jtagmkI.c.obj src\CMakeFiles\libavrdude.dir\jtagmkII.c.obj src\CMakeFiles\libavrdude.dir\jtag3.c.obj src\CMakeFiles\libavrdude.dir\linuxgpio.c.obj src\CMakeFiles\libavrdude.dir\linuxspi.c.obj src\CMakeFiles\libavrdude.dir\lists.c.obj src\CMakeFiles\libavrdude.dir\micronucleus.c.obj src\CMakeFiles\libavrdude.dir\par.c.obj src\CMakeFiles\libavrdude.dir\pgm.c.obj src\CMakeFiles\libavrdude.dir\pgm_type.c.obj src\CMakeFiles\libavrdude.dir\pickit2.c.obj src\CMakeFiles\libavrdude.dir\pindefs.c.obj src\CMakeFiles\libavrdude.dir\ppi.c.obj src\CMakeFiles\libavrdude.dir\ppiwin.c.obj src\CMakeFiles\libavrdude.dir\serbb_posix.c.obj src\CMakeFiles\libavrdude.dir\serbb_win32.c.obj src\CMakeFiles\libavrdude.dir\ser_avrdoper.c.obj src\CMakeFiles\libavrdude.dir\ser_posix.c.obj src\CMakeFiles\libavrdude.dir\ser_win32.c.obj src\CMakeFiles\libavrdude.dir\serialupdi.c.obj src\CMakeFiles\libavrdude.dir\stk500.c.obj src\CMakeFiles\libavrdude.dir\stk500v2.c.obj src\CMakeFiles\libavrdude.dir\stk500generic.c.obj src\CMakeFiles\libavrdude.dir\teensy.c.obj src\CMakeFiles\libavrdude.dir\updi_link.c.obj src\CMakeFiles\libavrdude.dir\updi_nvm.c.obj src\CMakeFiles\libavrdude.dir\updi_readwrite.c.obj src\CMakeFiles\libavrdude.dir\updi_state.c.obj src\CMakeFiles\libavrdude.dir\usbasp.c.obj src\CMakeFiles\libavrdude.dir\usb_hidapi.c.obj src\CMakeFiles\libavrdude.dir\usb_libusb.c.obj src\CMakeFiles\libavrdude.dir\usbtiny.c.obj src\CMakeFiles\libavrdude.dir\update.c.obj src\CMakeFiles\libavrdude.dir\wiring.c.obj src\CMakeFiles\libavrdude.dir\xbee.c.obj src\CMakeFiles\libavrdude.dir\__\lexer.c.obj src\CMakeFiles\libavrdude.dir\__\config_gram.c.obj /out:src\libavrdude.dll /implib:src\libavrdude.lib /pdb:src\libavrdude.pdb /dll /version:1.0 /machine:x64 /INCREMENTAL:NO _deps\libelf-build\libelf.lib _deps\libusb-build\libusb.lib C:\Program Files (x86)\Windows Kits\10\Lib\10.0.22000.0\um\x64\hid.lib _deps\libhidapi-build\libhidapi.lib _deps\libftdi-build\libftdi.lib setupapi.lib ws2_32.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:src\libavrdude.dll.manifest" failed (exit code 1120) with the following output:
     Creating library src\libavrdude.lib and object src\libavrdude.exp
C:\work\avr\avrdude\out\build\x64\usbasp.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\usbtiny.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\wiring.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\xbee.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\stk500.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\stk500v2.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\teensy.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\updi_link.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\ft245r.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\jtagmkII.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\micronucleus.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\pickit2.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\avrftdi_tpi.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\bitbang.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\buspirate.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\butterfly.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\arduino.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\avr.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\avr910.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\avrftdi.c.obj : error LNK2001: unresolved external symbol usleep
C:\work\avr\avrdude\out\build\x64\update.c.obj : error LNK2019: unresolved external symbol avrdude_message referenced in function dup_update
C:\work\avr\avrdude\out\build\x64\wiring.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\xbee.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\config_gram.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\usbasp.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\usb_hidapi.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\usb_libusb.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\usbtiny.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\teensy.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\updi_link.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\updi_nvm.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\updi_readwrite.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\serialupdi.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\stk500.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\stk500v2.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\stk500generic.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\pindefs.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\serbb_win32.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\ser_avrdoper.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\ser_win32.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\micronucleus.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\par.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\pgm.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\pickit2.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\jtagmkII.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\jtag3.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\linuxgpio.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\linuxspi.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\flip1.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\flip2.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\ft245r.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\jtagmkI.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\butterfly.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\config.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\dfu.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\fileio.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\avrftdi_tpi.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\avrpart.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\bitbang.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\buspirate.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\arduino.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\avr.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\avr910.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\avrftdi.c.obj : error LNK2001: unresolved external symbol avrdude_message
C:\work\avr\avrdude\out\build\x64\wiring.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\xbee.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\usb_hidapi.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\usb_libusb.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\usbtiny.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\update.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\updi_link.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\updi_nvm.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\updi_readwrite.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\usbasp.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\stk500.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\stk500v2.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\stk500generic.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\teensy.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\serbb_win32.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\ser_avrdoper.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\ser_win32.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\serialupdi.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\par.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\pgm.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\pickit2.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\pindefs.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\jtag3.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\linuxgpio.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\linuxspi.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\micronucleus.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\flip2.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\ft245r.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\jtagmkI.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\jtagmkII.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\config.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\dfu.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\fileio.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\flip1.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\avrpart.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\bitbang.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\buspirate.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\butterfly.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\arduino.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\avr.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\avr910.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\avrftdi.c.obj : error LNK2001: unresolved external symbol progname
C:\work\avr\avrdude\out\build\x64\updi_nvm.c.obj : error LNK2001: unresolved external symbol gettimeofday
C:\work\avr\avrdude\out\build\x64\xbee.c.obj : error LNK2001: unresolved external symbol gettimeofday
C:\work\avr\avrdude\out\build\x64\avr.c.obj : error LNK2001: unresolved external symbol gettimeofday
C:\work\avr\avrdude\out\build\x64\jtagmkII.c.obj : error LNK2001: unresolved external symbol gettimeofday
C:\work\avr\avrdude\out\build\x64\serialupdi.c.obj : error LNK2001: unresolved external symbol gettimeofday
C:\work\avr\avrdude\out\build\x64\stk500v2.c.obj : error LNK2001: unresolved external symbol gettimeofday
C:\work\avr\avrdude\out\build\x64\teensy.c.obj : error LNK2001: unresolved external symbol progbuf
C:\work\avr\avrdude\out\build\x64\usbasp.c.obj : error LNK2001: unresolved external symbol progbuf
C:\work\avr\avrdude\out\build\x64\usbtiny.c.obj : error LNK2001: unresolved external symbol progbuf
C:\work\avr\avrdude\out\build\x64\ft245r.c.obj : error LNK2001: unresolved external symbol progbuf
C:\work\avr\avrdude\out\build\x64\jtagmkII.c.obj : error LNK2001: unresolved external symbol progbuf
C:\work\avr\avrdude\out\build\x64\jtag3.c.obj : error LNK2001: unresolved external symbol progbuf
C:\work\avr\avrdude\out\build\x64\micronucleus.c.obj : error LNK2001: unresolved external symbol progbuf
C:\work\avr\avrdude\out\build\x64\avr.c.obj : error LNK2001: unresolved external symbol progbuf
C:\work\avr\avrdude\out\build\x64\fileio.c.obj : error LNK2001: unresolved external symbol progbuf
C:\work\avr\avrdude\out\build\x64\flip1.c.obj : error LNK2001: unresolved external symbol progbuf
C:\work\avr\avrdude\out\build\x64\flip2.c.obj : error LNK2001: unresolved external symbol progbuf
C:\work\avr\avrdude\out\build\x64\jtagmkII.c.obj : error LNK2019: unresolved external symbol ovsigck referenced in function jtagmkII_setparm
C:\work\avr\avrdude\out\build\x64\jtag3.c.obj : error LNK2001: unresolved external symbol ovsigck
C:\work\avr\avrdude\out\build\x64\serialupdi.c.obj : error LNK2001: unresolved external symbol ovsigck
C:\work\avr\avrdude\out\build\x64\avr910.c.obj : error LNK2001: unresolved external symbol ovsigck
C:\work\avr\avrdude\out\build\x64\flip1.c.obj : error LNK2001: unresolved external symbol ovsigck
C:\work\avr\avrdude\out\build\x64\flip2.c.obj : error LNK2001: unresolved external symbol ovsigck
C:\work\avr\avrdude\out\build\x64\ft245r.c.obj : error LNK2001: unresolved external symbol ovsigck
C:\work\avr\avrdude\out\build\x64\usbasp.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\usb_hidapi.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\usb_libusb.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\usbtiny.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\jtagmkII.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\jtag3.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\ser_avrdoper.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\ser_win32.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\flip1.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\flip2.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\ft245r.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\jtagmkI.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\avrftdi.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\bitbang.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\butterfly.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\dfu.c.obj : error LNK2001: unresolved external symbol verbose
C:\work\avr\avrdude\out\build\x64\fileio.c.obj : error LNK2019: unresolved external symbol quell_progress referenced in function fileio
C:\work\avr\avrdude\out\build\x64\stk500v2.c.obj : error LNK2001: unresolved external symbol quell_progress
C:\work\avr\avrdude\out\build\x64\usbtiny.c.obj : error LNK2001: unresolved external symbol quell_progress
C:\work\avr\avrdude\out\build\x64\update.c.obj : error LNK2001: unresolved external symbol quell_progress
C:\work\avr\avrdude\out\build\x64\src\libavrdude.dll : fatal error LNK1120: 8 unresolved externals
  ninja: build stopped: subcommand failed.

Build All failed.
```

</details>

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -118,6 +118,7 @@ if(WIN32)
     set(PREFERRED_LIBHIDAPI libhidapi.a libhidapi-libusb.a libhidapi-hidraw.a hidapi hidapi-libusb hidapi-hidraw)
     set(PREFERRED_LIBFTDI libftdi.a ftdi)
     set(PREFERRED_LIBFTDI1 libftdi1.a ftdi1)
+    set(PREFERRED_LIBREADLINE libreadline.a)
 else()
     set(PREFERRED_LIBELF elf)
     set(PREFERRED_LIBUSB usb)
@@ -125,6 +126,7 @@ else()
     set(PREFERRED_LIBHIDAPI hidapi hidapi-libusb hidapi-hidraw)
     set(PREFERRED_LIBFTDI ftdi)
     set(PREFERRED_LIBFTDI1 ftdi1)
+    set(PREFERRED_LIBREADLINE readline)
 endif()
 
 # -------------------------------------
@@ -212,7 +214,7 @@ endif()
 # -------------------------------------
 # Find libreadline
 
-find_library(HAVE_LIBREADLINE NAMES readline)
+find_library(HAVE_LIBREADLINE NAMES ${PREFERRED_LIBREADLINE})
 if(HAVE_LIBREADLINE)
     set(LIB_LIBREADLINE ${HAVE_LIBREADLINE})
 endif()

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -35,6 +35,13 @@ option(USE_LIBUSBWIN32 "Prefer libusb-win32 over libusb" OFF)
 option(DEBUG_CMAKE "Enable debugging output for this CMake project" OFF)
 option(BUILD_SHARED_LIBS "Build shared libraries" OFF)
 
+if(WIN32)
+    # Prefer static libraries over DLLs on Windows
+    option(USE_STATIC_LIBS "Use static libraries" ON)
+else()
+    option(USE_STATIC_LIBS "Use static libraries" OFF)
+endif()
+
 include(CheckIncludeFile)
 include(CheckSymbolExists)
 include(FetchContent)
@@ -110,8 +117,7 @@ endif()
 # Detect installed libraries
 # =====================================
 
-# Prefer static libraries over DLLs on Windows
-if(WIN32)
+if(USE_STATIC_LIBS)
     set(PREFERRED_LIBELF libelf.a elf)
     set(PREFERRED_LIBUSB libusb.a usb)
     set(PREFERRED_LIBUSB_1_0 libusb-1.0.a usb-1.0)

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -51,7 +51,7 @@ include_directories(BEFORE ${CMAKE_CURRENT_BINARY_DIR})
 add_compile_definitions(CONFIG_DIR=\"${CONFIG_DIR}\")
 
 if(WIN32)
-    set(EXTRA_WINDOWS_SOURCES "${PROJECT_BINARY_DIR}/src/windows.rc")
+    set(EXTRA_WINDOWS_RESOURCES "${PROJECT_BINARY_DIR}/src/windows.rc")
     set(EXTRA_WINDOWS_LIBRARIES setupapi ws2_32)
 endif()
 
@@ -216,6 +216,7 @@ add_library(libavrdude
     xbee.c
     ${FLEX_Parser_OUTPUTS}
     ${BISON_Parser_OUTPUTS}
+    "${EXTRA_WINDOWS_SOURCES}"
     )
     
 set_target_properties(libavrdude PROPERTIES
@@ -253,7 +254,7 @@ add_executable(avrdude
     term.h
     whereami.c
     whereami.h
-    "${EXTRA_WINDOWS_SOURCES}"
+    "${EXTRA_WINDOWS_RESOURCES}"
     )
 
 target_link_libraries(avrdude PUBLIC libavrdude)

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -55,11 +55,6 @@ if(WIN32)
     set(EXTRA_WINDOWS_LIBRARIES setupapi ws2_32)
 endif()
 
-if(NOT WIN32)
-    set(LIB_MATH m)
-    add_compile_options(-Wall) # -Wextra
-endif()
-
 if(MSVC)
     add_compile_definitions(_CRT_SECURE_NO_WARNINGS=1)
     add_compile_definitions(_CRT_NONSTDC_NO_WARNINGS=1)
@@ -79,6 +74,9 @@ if(MSVC)
     set(EXTRA_WINDOWS_INCLUDES ${EXTRA_WINDOWS_INCLUDES}
         "msvc"
         )
+else()
+    set(LIB_MATH m)
+    add_compile_options(-Wall) # -Wextra
 endif()
 
 # =====================================

